Convert pic24 build system to use scons and make momo tool usable

timburke opened the pull request:
This PR replaces rake with an scons based build system for pic24.  All of the pic24 modules should compile now with the new build system.  There are still two outstanding issues:
- [ ] unit testing is not yet complete.  I built a simulation framework to handle the sim30 and gpsim programs in a generic way but it's not yet finished.  
- [ ] Compiling an application on pic24 does not check for changes to shared files and recompile if the library has changed.  This requires implementing better dependency scanning, which is well-supported by scons but not yet implemented into the build system.
### Build System Usage

To build pic24 modules, use the same command as for pic12 projects:

``` shell
$ momo build
```
### MoMo tool

I have added a shell to the MoMo tool and am including it in this PR mainly because it will be useful for demos in Spain.  Basically, the momo tool takes command line arguments and maps them to python functions via decorator annotations on the functions so that it knows how to convert the parameters to the appropriate types.  If the return value from the function is a context type object, i.e. it has annotated routines that can be run, this context becomes the scope of the next call.  If the command line passed to MoMo is either empty of ends in a context object, a shell is opened and commands are allowed until 'quit' is entered or ctrl-D is sent to indicate and EOF.

Example help usage:

``` shell
$ momo
(root) help
root

Defined Functions:
 - controller(string serial='')
 - HexFile(path source, integer wordsize, integer filewords, integer skip=1, integer cutoff=65535)

Builtin Functions
 - quit: quit the momo shell
 - help [function]: print help information about the current context or a function
 - back: replace the current context with its parent

(root) 
```

The same python annotations are used to build help documentation for all of the functions you can call from momo.

timburke committed:
Rename ASM file so .o output does not conflict with c file.

timburke committed:
Add tool support for scons builds of pic24 code

- Add xc16 Builder objects (cc, ld, ar)
- Add generic support for storing per chip family includes and defines
- Add support for having chip specific defines overriden by module
specific defines
- Add easy autobuild support
- Add deprecated decorator for telling where API functions are used

timburke committed:
Pull in correct xc16 header based on target processor

timburke committed:
Add SConstruct build files for library, controller and bootloader

timburke committed:
Fix mibtool to always know where the site_scons folder is

timburke committed:
Add scons build files for field_service_unit

timburke committed:
Fix bug when the default value of the property should be None

timburke committed:
Make shared library building modular

timburke committed:
Completely remove ceedling from tools

timburke committed:
Finish removing all traces of ceedling from repo.

timburke committed:
Merge branch 'pr-executive-upgrade' into feature-pic24-build

timburke committed:
Expand typedargs to include path type

Paths are strings that specify files.  Validators exist for
specifying that the file must exist, be readable and be writable

timburke committed:
Add setting for sim30 path to support simulating pic24

timburke committed:
First version of infrastructure for generic PIC simulation

Simulator class is a simulator agnostic interface to a PIC simulator
that lets you control the simulator completely.  It spawns a
separate thread to run the simulator in for UI responsiveness.

Concrete simulators are implemented by subclassing SimWorker. Current
basic implementation exists for sim30 to simulate PIC24 code. This
will be used for pic24 unit testing and then pic12 unit tests will
be ported to use this new interface.

timburke committed:
Update typedargs to always throw only MoMoException subclasses

We should standardize the exceptions that API routines throw so
that callers know what to expect and what means what.

timburke committed:
Change exceptions to log strings correctly

Previously would log a tuple with one string inside

timburke committed:
Add asynchronous file reading module

It's surprisingly hard to read from a pipe asynchronously in python
in a cross-platform way.  select() and poll() do not work on Windows
when given a pipe or file object.  The only way that seems to work
on all platforms is spinning up a dedicated thread to block reading
the file and then communicate with that thread using a threadsafe
Queue.

This is not efficient for large reading but is fine for reading a
few bytes.  This is needed for safe subprocess communication with
external simulator programs.

timburke committed:
Update sim30 module to use asyncio for reading from sim30 program.
This is to be able to test when the simulator is successfully
running code and when not.  Reads will timeout when it is running
code since it seems to be single threaded.

timburke committed:
Add peek to asyncio

timburke committed:
Add ability to wait for execution to finish

timburke committed:
Fix exception type

timburke committed:
Add hex package to manage both pic12 and pic24 hex files

Provides a generic and unified API for the pic12 and pic24 address
space.  Will eventually obsolete the hex8 and hex16 packages

timburke committed:
Update momo tool to include a shell and print better help

If you pass a command line to momo that ends not with data but with
a context object, you will be presented with a shell that starts
with that context.  The help system now presents much better help
documentation as well, pulled from the context and function docstrings

timburke committed:
Update shell with new builtins, fix parameter handling

- can now navigate up the context chain
- new ArgumentError for invalid arguments
- better handling of combined kw and positional arguments

timburke committed:
Fix argument exception message to be less repetitive

timburke commented:
closes #95

timburke committed:
Update momo to be able to build pic24 and pic12 projects

timburke committed:
Update momo tool to allow for parallel builds

You cannot run scons via the import command in python. For whatever
reason, it causes the command to hang when you do a parallel build
with -jN.  Running via a simple function wall works.

timburke committed:
Add timeout to wait_prompt

timburke committed:
Change method from quit to exit to not conflict with momo shell builtin

timburke committed:
Change asyncio to be line oriented with a configuratable separator string

timburke committed:
Add ability to run programs, wait for them and parse errors to SIM30

timburke committed:
Update shell with common error handling

timburke committed:
Update momo to latest version

timburke committed:
Add ability to attach logs to simulator output

- currently attaches to UART1 in sim30
- builds a temporary file that is automatically disposed upon cleanup
- contents are available as a string by calling get_log after calling
attach_log, running the simulator and then calling finish

timburke committed:
Update simulator and build system.

timburke committed:
Rewrite ChipFamily and ChipSettings to be more general

New build system allows for an arbitrary number of architectures
to be specified for each module.  Architectures can override definitions
from archs earlier in the list and modules can define overlays for
each architecture.  A chip is just a specific architecture.  This will
allow for more easy expansion to building unit tests on pic24 since
you will be able to add the test architecture on top of all other
architectures that you build for and test each one.

timburke committed:
Update PIC24 Build System

- Hide command line by default and print summary message
- Use TargetSettings object to contain all of the build information
- Properly name modules and architectures in a uniform fashion

timburke committed:
Python unit tests fror pic24 build system

Uses pyunit and nose for testing.  Current tests:
- build_settings file exists and is valid
- ChipFamily and TargetSettings objects work

timburke committed:
Merge branch 'feature-pic24-build' of github.com:timburke/MoMo-Firmware into feature-pic24-build

timburke committed:
Finish updating build_settings.json for compat with new build system

timburke committed:
Finish updating xc8 build system to use new TargetSettings interface.

- Unit testing now works again on xc8
- Both systems use the new ChipFamily/TargetSettings interface for
managing build configuration information

timburke committed:
Update unit test for typo

timburke committed:
Update pymomo unit tests for latest build system changes

timburke committed:
Fix incorrect import

timburke committed:
Add ability to produce retargeted modules for compilation

This simplifies the creation of unit tests on pic12 and will be
very helpful for pic24 as well.

timburke committed:
Add patch for latest svn HEAD of gpsim to fix it.

There are errors in the current head of gpsim that prevent it from
correctly simulating Enhanced Midrange PICs in certain subtle ways
that break unit tests for MoMo.  This patch applies cleanly to
gpsim svn HEAD and fixes the issues.  A bug report has been submitted
with the patch so hopefully this will be fixed upstream in the future.

timburke committed:
Use utilities version of function

timburke committed:
Update gpsim unit test code to work with latest gpsim.

Now we no longer need a custom version of gpsim but can work directly
from the gpsim svn repo with a simple patch.  Setting this up will
be documented on the wiki.

timburke committed:
Fix quoting bug with spaces in path

timburke committed:
Working pic24 unit test construction

TODO:
- implement main function
- write runner for simulator
- write log file parser

Finished:
- Unit test type is now autoselected based on the type specified in
the file

timburke committed:
Update shared unit tests on pic24 to use new unit test format

timburke committed:
Change error type for consistency

timburke committed:
Add Unity test framework to the repository

timburke committed:
Merge branch 'dev' of github.com:WellDone/MoMo-Firmware into feature-pic24-build

Conflicts:
	config/build_settings.json

timburke committed:
Merge branch 'dev' of github.com:WellDone/MoMo-Firmware into feature-pic24-build

timburke committed:
Add ability to specify more xc16 compiler options

timburke committed:
Update build_settings.json

timburke committed:
Finish tool support for PIC24 unit testing

timburke committed:
Update unity to work with PIC24FJ64GA306 and PIC24F16KA101

timburke committed:
Add nice string for when compiler is running

timburke committed:
Add ability to specify model of pic24 to be simulated

24FJ64GA306 required mode 'pic24fpsuper', 24F16KA101 requires 'pic24super'

timburke committed:
Start adding support for running unit tests

timburke committed:
Add initial unit testing support for simulator infrastructure

timburke committed:
Add linker files for simulating pic24 procs

timburke committed:
Update linkedlist test to work with new unit testing framework

timburke committed:
Update RTCC module to compile on old PIC24 processor

A register was added on the PIC24FJ64GA306 that's not present on the
older proc, so add an ifdef to ignore that register on the older chip

timburke commented:
The build system should be ready to go at this point.  To build any module now you should use momo build instead of mibtool build.  Usage:

``` shell
> cd momo_modules/shared/pic24
> momo build
```

@amcgee This PR is ready for review.  The unit testing support is not complete yet though.  I have the pic24 simulator working and module unit tests which are unit tests that just test a single .c,.h module will compile and run in the simulator, however, I have not automated the running process nor collected the results to build a summary like I did on pic12.

timburke commented:
Tracking unit testing progress in Issue #117.

amcgee commented on config/build_settings.json:
```diff
@@ -16,86 +12,92 @@
-			"gsm_module": ["16lf1823"],
-			"pv_module": ["16lf1847"],
-			"multi_sensor_module": ["16lf1847"]
+			"gsm_module": {},
```
I'd really like to avoid global module definition requirements as much as possible, can't these be defined in the SConscript file local to each module?  These global dependencies already prevent splitting repositories, so if this is a big undertaking we can defer the change and do it with that effort, otherwise we should avoid adding this complexity.

timburke replied:
Interesting.  I agree with the thinking here.  I can move the per module settings into each module's folder in a separate configuration file.  There are some settings that are global to a chip family though.  For example, on pic12, it's really important that everybody have the same definitions for what rom page the application modules start on, what MIB version to say they were compiled for etc.  How would you propose we handle those types of settings?

timburke replied:
It's straightforward to move the module definitions into each module's folder.  I can take care of that.

amcgee commented on config/settings.global.json:
```diff
@@ -2,6 +2,7 @@
 		"eagle": "REPLACE WITH PATH TO EAGLE",
-		"gpsim": "REPLACE WITH PATH TO GPSIM IF YOU WANT TO SIMULATE PIC12 AND PIC16 (REQUIRED FOR UNIT TESTS)"
+		"gpsim": "REPLACE WITH PATH TO GPSIM IF YOU WANT TO SIMULATE PIC12 AND PIC16 (REQUIRED FOR UNIT TESTS)",
+		"sim30": "REPLACE WITH PATH TO SIM30 IF YOU WANT TO SIMULATE PIC24 (REQUIRED FOR UNIT TESTS)"
```
These should probably be environment variables and should be set automatically in the development environment (Vagrant) provisioning scripts.

timburke replied:
I agree with this.  How would you feel about having a strategy of first looking for an environment variable if one is set and second checking if the executable is in the PATH?  If neither of the two works then we could raise an exception.  That would allow for vagrant to set up the environment variables correctly but also mean that people wouldn't have to pollute their environment variables with lots of random definitations if they didn't want to use vagrant

amcgee replied:
PATH is fine for Vagrant as well, that's definitely the way to go.  No need to support custom environment variables as well I don't think.

amcgee commented on tools/site_scons/utilities.py:
```diff
@@ -26,250 +28,53 @@ def find_files(dirname, pattern):
-def for_all_targets(module, func):
-	targets = get_module_targets(module)
+	#Append chip type and suffix
+	libs = ["%s_%s" % (x, chip.arch_name()) for x in libs]
```
On my system this lib (pic24_24fj64ga306) isn't found.  Either there's an undocumented setup step or the name is wrong - I do have a libpPIC24Fxxx-elf.a file in my XC16 1.21 install that might be it?

timburke replied:
I'll look into this.  That library should be the output of building in momo_modules/shared/pic24 since it's the shared library with all the pic24 modules.  Undocumented setup step.  Does

``` shell
$ cd momo_modules/shared/pic24
$ momo build
```

before building any of the pic24 application modules fix the problem?  Otherwise it's probably a path issue.  I'll look into it.

amcgee replied:
Ah, that might be it.  I assumed building mainboard would also build shared.  Is it possible to automatically track dependencies and build them?

amcgee commented:
Mostly looks good, I noted a few issues that came up when testing.

Also, the README files for each PIC24 module (mainboard and mainboard_bootloader) should be updated or removed, and the setup scripts for the Vagrant dev environment need updating.

timburke committed:
Move module settings to per module file

timburke committed:
Update build system to support module settings files

This allows us to have module setting stored locally with each module

timburke committed:
Update SConstruct files to include local module settings

timburke committed:
Update pic24 library to be slimmer on pic24f16ka101

Current version include many mainboard specific routines like
system_log, flash_queue and flash_block that require a flash memory
chip.  The FSU does not have one so it cannot use these routines.

The routines were making the code size too big to fit on the chip.

timburke committed:
Update to look for external programs in the PATH

timburke committed:
Update to not pull in pic24 feature definitions

This is never referenced in the PIC12 executive code, but xc8 does
not seem capable of optimizing it out (but only on 16lf1847) so it
causes problems with the memory layout.

timburke committed:
Remove out of date README from mainboard

timburke commented:
Current status of this is:
- module settings stored per module rather than in global config
- external tools searched for in PATH
- fixed compilation of fsu firmware
- fixed compilation of pic12_executive units tests

amcgee commented:
Looking better, now I'm getting a python error building pic24 shared:

```
vagrant@precise64:/vagrant/momo_modules/shared/pic24$ momo build
scons: Reading SConscript files ...
AttributeError: UnitTest instance has no attribute 'type':
  File "/vagrant/momo_modules/shared/pic24/SConstruct", line 16:
    unit_test.build_units('test', family.targets(module_name))
  File "/vagrant/tools/site_scons/unit_test.py", line 244:
    tests = find_units(parent)
  File "/vagrant/tools/site_scons/unit_test.py", line 213:
    if unit.type not in known_types:
```

The known_types dictionary is empty, so even if the .type property did exist the test would never return true.

amcgee commented:
I take that back, it looks like it's defined externally in unit_test12 and unit_test24.  Doing this kind of global modification is bad practice, the known_types should probably be an instance member of the UnitTest class and passed in through the constructor there.

Regardless, UnitTest.type isn't being defined.

timburke commented:
I think that is probably because there are not proper headers on two of 
the unit tests in shared/pic24/test.  I had removed them from my local 
copy while I got the unit testing working.  I can fix this but if you 
look at test_linkedlist.c you'll see a header in the comments section.  
One of the attributes it defines is whether the test is a module test, a 
whole application test or an executive test (latter two apply only to 
pic12).  So it's filled in from the header it reads from the unit test.
- tim

On 9/30/14, 12:33 PM, Austin McGee wrote:

> Looking better, now I'm getting a python error building pic24 shared:
> 
> |vagrant@precise64:/vagrant/momo_modules/shared/pic24$ momo build
> scons: Reading SConscript files ...
> AttributeError: UnitTest instance has no attribute 'type':
>    File "/vagrant/momo_modules/shared/pic24/SConstruct", line 16:
>      unit_test.build_units('test', family.targets(module_name))
>    File "/vagrant/tools/site_scons/unit_test.py", line 244:
>      tests = find_units(parent)
>    File "/vagrant/tools/site_scons/unit_test.py", line 213:
>      if unit.type not in known_types:
> |
> 
> The known_types dictionary is empty, so even if the .type property did 
> exist the test would never return true.
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/WellDone/MoMo-Firmware/pull/105#issuecomment-57369432.

amcgee commented:
Ok, if that's the case the error message should be less cryptic.  Any headers that are required should throw an explicit error in _extract_header when they don't exist.

amcgee commented:
If unit tests aren't supported yet on pic24 let's not run them by default.

amcgee commented:
A few more things I've noticed...

There's no longer an option to clean up build artifacts or force a rebuild.  Previously rake would rebuild automatically, not it does an incremental build and there's no `momo build clean` or similar option.

If you call the momo tool with a non-existent command, it opens the shell.  It should exit with a non-zero exit code.  I.E.  try running `momo foobar`

timburke commented:
Word. I can fix the shell opening.  momo build just internally calls 
scons so you can clean all build artifacts with
momo build -c
- tim
  On 9/30/14, 4:19 PM, Austin McGee wrote:
  >
  > A few more things I've noticed...
  >
  > There's no longer an option to clean up build artifacts or force a 
  > rebuild. Previously rake would rebuild automatically, not it does an 
  > incremental build and there's no |momo build clean| or similar option.
  >
  > If you call the momo tool with a non-existent command, it opens the 
  > shell. It should exit with a non-zero exit code. I.E. try running 
  > |momo foobar|
  >
  > —
  > Reply to this email directly or view it on GitHub 
  > https://github.com/WellDone/MoMo-Firmware/pull/105#issuecomment-57397984.
  >

amcgee commented:
Ok cool, that should probably be documented somewhere (momo help build) says there are no available arguments.

amcgee commented:
I see that it does say "exactly as if the scons tool had been invoked", guess it's just a bit ambiguous.

timburke commented:
I agree.  This comment chain will serve as my todo list on this.

On 9/30/14, 4:22 PM, Austin McGee wrote:

> Ok cool, that should probably be documented somewhere (momo help 
> build) says there are no available arguments.
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/WellDone/MoMo-Firmware/pull/105#issuecomment-57398290.

amcgee commented:
Great.  I'm setting up automatic CI testing using Travis and this new system.

amcgee commented:
This is blocking a bunch of stuff so I'm going to just merge it as-is.  We still need to address the relatively minor issues remaining.
